Warning about assert None from XFAIL tests

Running pytest master I see warnings from XFAIL tests about asserting None. The warnings suggests to change the assert to `assert obj is None` but the reason we are asserting `None` is because the test fails (hence `XFAIL`).

I don't think these warnings should be emitted from XFAIL tests. 

Using pytest master:
```python
import pytest

@pytest.mark.XFAIL
def test_f():
    assert g()

def g():
    # Shouldn't return None but doesn't work properly...
    return None
```

Running the tests gives:
```console
$ pytest test_f.py 
====================================================== test session starts =======================================================
platform darwin -- Python 2.7.10, pytest-4.1.1, py-1.7.0, pluggy-0.8.0
rootdir: /Users/enojb/current/sympy/pytest, inifile: tox.ini
plugins: xdist-1.26.0, forked-0.2, doctestplus-0.3.0.dev0
collected 1 item                                                                                                                 

test_f.py F                                                                                                                [100%]

============================================================ FAILURES ============================================================
_____________________________________________________________ test_f _____________________________________________________________

    @pytest.mark.XFAIL
    def test_f():
>       assert g()
E       PytestWarning: asserting the value None, please use "assert is None"

test_f.py:5: PytestWarning
==================================================== short test summary info =====================================================
FAIL test_f.py::test_f
==================================================== 1 failed in 0.08 seconds ====================================================
```

I believe the mistake here is `pytest.mark.XFAIL` vs `pytest.mark.xfail`

trying in a little script this works as expected, even with `assert None`:

### `pytest.mark.xfail`

```console
$ cat t.py
import pytest

@pytest.mark.xfail
def test():
    assert None
$ pytest t.py 
============================= test session starts ==============================
platform linux -- Python 3.6.7, pytest-4.1.1, py-1.7.0, pluggy-0.8.1
rootdir: /tmp/t, inifile:
collected 1 item                                                               

t.py x                                                                   [100%]

=============================== warnings summary ===============================
t.py::test
  /tmp/t/t.py:5: PytestWarning: asserting the value None, please use "assert is None"
    assert None

-- Docs: https://docs.pytest.org/en/latest/warnings.html
==================== 1 xfailed, 1 warnings in 0.07 seconds =====================
$ pytest -Werror t.py 
============================= test session starts ==============================
platform linux -- Python 3.6.7, pytest-4.1.1, py-1.7.0, pluggy-0.8.1
rootdir: /tmp/t, inifile:
collected 1 item                                                               

t.py x                                                                   [100%]

========================== 1 xfailed in 0.03 seconds ===========================
```

### `pytest.mark.XFAIL`

switching to the incorrect name:

```console
$ cat t.py
import pytest

@pytest.mark.XFAIL
def test():
    assert None
$ pytest t.py
============================= test session starts ==============================
platform linux -- Python 3.6.7, pytest-4.1.1, py-1.7.0, pluggy-0.8.1
rootdir: /tmp/t, inifile:
collected 1 item                                                               

t.py F                                                                   [100%]

=================================== FAILURES ===================================
_____________________________________ test _____________________________________

    @pytest.mark.XFAIL
    def test():
>       assert None
E       assert None

t.py:5: AssertionError
=============================== warnings summary ===============================
t.py::test
  /tmp/t/t.py:5: PytestWarning: asserting the value None, please use "assert is None"
    assert None

-- Docs: https://docs.pytest.org/en/latest/warnings.html
===================== 1 failed, 1 warnings in 0.05 seconds =====================
$ pytest -Werror t.py
============================= test session starts ==============================
platform linux -- Python 3.6.7, pytest-4.1.1, py-1.7.0, pluggy-0.8.1
rootdir: /tmp/t, inifile:
collected 1 item                                                               

t.py F                                                                   [100%]

=================================== FAILURES ===================================
_____________________________________ test _____________________________________

    @pytest.mark.XFAIL
    def test():
>       assert None
E       _pytest.warning_types.PytestWarning: asserting the value None, please use "assert is None"

t.py:5: PytestWarning
=========================== 1 failed in 0.04 seconds ===========================
```

@oscarbenjamin 
> the error message itself could be improved.

Do you have a suggestion?
Should we use a more verbose error message with "assert None" instead, e.g.

```
>       assert os.remove.assert_has_calls(calls, any_order=True)
E       WARNING: asserting the value None, please use "assert is None"
E       AssertionError: assert None
E        +  where None = <bound method wrap_assert_has_calls of <MagicMock name='remove' id='65891856'>>([call('file_0.txt'), call('file_1.txt'), call('file_2.txt'), call('file_3.txt'), call('file_4.txt'), call('file_5.txt'), ...], any_order=True)
E        +    …

I agree it should be removed, because we can't distinguish a valid `None` value from "accidentally calling `assert` with a function which doesn't `return` at all".

Recently on Twitter I saw a similar case which is tricky but not caught by this warning:

```python
def test_values():
    a = [1, 2, 3]
    b = [3, 2, 1, 4]
    assert a.sort() == b.sort()
```

The test silently passes, but it's not immediately visible why. The real solution, IMHO, is starting to run `mypy` with `--check-untyped-defs` over the testsuite - because mypy *can* known whether a function does return a value or doesn't. In this case, it shows:

```
test_sort.py:4: error: "sort" of "list" does not return a value
```

Unfortunately, it doesn't catch something like the original [example](https://stackoverflow.com/q/48152560) mentioned by @RonnyPfannschmidt in #3191:

```python
import unittest.mock

def test_mock():
    m = unittest.mock.Mock()
    m.meth(123)
    assert m.meth.assert_called_with(123)
```

But still, I think mypy is a much more suitable place for such a check than pytest is, because it does have the distinction between something like `-> None` and `-> Optional[re.Match]`.

--- a/changelog/4639.feature.rst
+++ b/changelog/4639.feature.rst
@@ -1,0 +1,3 @@
+Revert "A warning is now issued when assertions are made for ``None``".
+The warning proved to be less useful than initially expected and had quite a
+few false positive cases.

--- a/src/_pytest/assertion/rewrite.py
+++ b/src/_pytest/assertion/rewrite.py
@@ -799,13 +799,6 @@ class AssertionRewriter(ast.NodeVisitor):
         self.push_format_context()
         # Rewrite assert into a bunch of statements.
         top_condition, explanation = self.visit(assert_.test)
-        # If in a test module, check if directly asserting None, in order to warn [Issue #3191]
-        if self.module_path is not None:
-            self.statements.append(
-                self.warn_about_none_ast(
-                    top_condition, module_path=self.module_path, lineno=assert_.lineno
-                )
-            )
 
         negation = ast.UnaryOp(ast.Not(), top_condition)
 
@@ -886,30 +879,6 @@ class AssertionRewriter(ast.NodeVisitor):
         for stmt in self.statements:
             set_location(stmt, assert_.lineno, assert_.col_offset)
         return self.statements
-
-    def warn_about_none_ast(self, node, module_path, lineno):
-        """
-        Returns an AST issuing a warning if the value of node is `None`.
-        This is used to warn the user when asserting a function that asserts
-        internally already.
-        See issue #3191 for more details.
-        """
-        val_is_none = ast.Compare(node, [ast.Is()], [ast.NameConstant(None)])
-        send_warning = ast.parse(
-            """\
-from _pytest.warning_types import PytestAssertRewriteWarning
-from warnings import warn_explicit
-warn_explicit(
-    PytestAssertRewriteWarning('asserting the value None, please use "assert is None"'),
-    category=None,
-    filename={filename!r},
-    lineno={lineno},
-)
-            """.format(
-                filename=fspath(module_path), lineno=lineno
-            )
-        ).body
-        return ast.If(val_is_none, send_warning, [])
 
     def visit_Name(self, name):
         # Display the repr of the name if it's a local variable or

--- a/testing/test_warnings.py
+++ b/testing/test_warnings.py
@@ -543,7 +543,7 @@ class TestAssertionWarnings:
 
     def test_tuple_warning(self, testdir):
         testdir.makepyfile(
-            """
+            """\
             def test_foo():
                 assert (1,2)
             """
@@ -552,48 +552,6 @@ class TestAssertionWarnings:
         self.assert_result_warns(
             result, "assertion is always true, perhaps remove parentheses?"
         )
-
-    @staticmethod
-    def create_file(testdir, return_none):
-        testdir.makepyfile(
-            """
-            def foo(return_none):
-                if return_none:
-                    return None
-                else:
-                    return False
-
-            def test_foo():
-                assert foo({return_none})
-            """.format(
-                return_none=return_none
-            )
-        )
-
-    def test_none_function_warns(self, testdir):
-        self.create_file(testdir, True)
-        result = testdir.runpytest()
-        self.assert_result_warns(
-            result, 'asserting the value None, please use "assert is None"'
-        )
-
-    def test_assert_is_none_no_warn(self, testdir):
-        testdir.makepyfile(
-            """
-            def foo():
-                return None
-
-            def test_foo():
-                assert foo() is None
-            """
-        )
-        result = testdir.runpytest()
-        result.stdout.fnmatch_lines(["*1 passed in*"])
-
-    def test_false_function_no_warn(self, testdir):
-        self.create_file(testdir, False)
-        result = testdir.runpytest()
-        result.stdout.fnmatch_lines(["*1 failed in*"])
 
 
 def test_warnings_checker_twice():

